refactor distribution detection in facts.py

robinro opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### ANSIBLE VERSION

devel
##### SUMMARY

This commit refactors the distribution version parsing in `module_utils/facts.py`. By doing so it fixes #15087. 

The old version had many duplications and got up to 12 levels of indentation. The refactor splits off methods for each distribution, reduces duplication and aims to improve readability.

I tried to keep the original logic as far as possible. The logic is rather weird in some places, which I now marked 'TODO', but don't want to change yet without being able to test them.
The overall idea is that a list of 'path, name' is iterated and `path` is read to get a distribution_version for `name` out of it. Whether or not the parsing is successful is not always obvious and sometimes the code assumes that even after parsing one file other files are still read. 

I also added a test_distribution_version.py that should be extended to test the correct parsing of `/etc/*release` files of different distributions. The tests are non-trivial as also calls to the module `platform` have to be mocked.

robinro commented:
The overall setup should probably be adjusted. At the moment I moved everything related to the distribution-guessing to a class `Distribution`. I don't yet see how to integrate this in the bigger picture behind the `FACT_SUBSETS` that @bcoca introduced recently, but I guess factoring out the distribution code as I did is a first step to any way of making `facts` more modular.

bcoca commented on lib/ansible/module_utils/facts.py:
```diff
@@ -157,32 +143,32 @@ class Facts(object):
         # instance, os_family and distribution).  Have to sort out what to do
         # about those first.
-        if load_on_init:
-            self.get_platform_facts()
```
don't remove this `if`

robinro replied:
readded it. I didn't see it used when I touched that part, maybe because the branchpoint was too old before the rebase.

bcoca commented:
not sure that the change makes this better, specially the restructuring of the family distro lookup is now obscured and generated from the hardcoded lists when before it was just hardcoded itself.

robinro commented:
The OS_FAMILY change is not really core to the refactor, so I reverted it. I found the old way of writing it very hard to read, but that might be just taste.

The main point is splitting up the `get_distribution_facts` method and making the individual parts more consistent.

@bcoca can you have a look at the `test_distribution_version.py`?
This should in principle also work with the old `get_distribution_version` implementation, but I'm unsure whether my use of mocking and calling the whole Facts class is still an appropriate use of the tests. If it is I would try to add more examples to get this as close to comprehensive as possible.

The main motivation for the refactor is the rather often occuring bugs with seldomly used distros (SuSE in my case), where some corner-case is lost due to the non-obvious code.

robinro commented:
The changes are now rebased/squashed to clean up the mess I went through to make Travis happy.

The test set is already quite long: Ubuntu 12.04, 14.04, Debian wheezy, jessie, stretch, openSuSE, SLES, CoreOS, ...
Independent of the refactor the test should really make it into devel. The test and facts.py is already python3 ready, but depends on `basic.py` which is not.

I had a look at the code to see whether more intrusive refactoring might be an option. I believe for the moment without more test coverage it's not. The distribution information is used in Facts(), so this has to be cleaned up first. Also the current setup depends on the order in which things are read. So I tried to split it up in small pieces and put some `TODO`s on the obscure bits. When we have test cases for them one could try a furthergoing refactor.

alikins commented:
The Distribution() class here looks good to me. A couple of thoughts:

(for discussion, not blocking...)
    \* It might be useful to pull all the mapping data out into a config/data file (OS_FAMILY, OSDIST_LIST, etc). 
        \* I wonder if there is a good upstream source for that info? Seems like it could be useful for many tools. If not, it may be useful to start one   
- A tiny potential general case optimization would be to sort OSDIST by popularity. Only suggested since the current list doesn't appear to have any ordering. I'd prefer that list be either sorted alphabetically for readability, or in rough order of commonness as mentioned above.

robinro commented:
@alikins Thanks for your comments
1. I'm not sure about having an external config; if doing so adds more lines than the data is long it's not worth the effort; it's probably more straight forward to make the whole thing a separate python module
2. There is some upstream I know of doing this, the cpython stdlib:
   https://github.com/python/cpython/blob/master/Lib/platform.py
   But it's a mess. The code in ansible is (in theory) only those parts that they get wrong. Their update cycle is quite slow and the coverage of distribution also not ideal.
   I also thought about making the whole thing a standalone python module. Maybe sometime...
3. Sorting OSDIST_LIST is dangerous, as some parts depend on the ordering (i.e. SuSE). If all the open TODOs in the code are fixed it should be ordering-independent. But at least the SuSE code needs some work for this.

robinro commented:
As discussed in https://meetbot.fedoraproject.org/teams/ansible_core_meeting/ansible_core_meeting.2016-04-14-15.50.log.html this is moved to 2.2.0. It should be included eary in the cycle.

abadger commented:
Okay, stable-2.1 has branched so we're ready to merge things to devel (2.2) now!  Feel free to rebase, cleanup merge commits, squash if you like, and then ping me on irc to merge this!

Commit:
refactor distribution version parsing

* split code as separate class
* split different distributions as individual functions
* keep program logic mostly identical (for now)
* increase readability, reduce complexity/indentation
* make future testing/refactoring easier
* step towards making distribution parsing independent of the Facts class
* add some changes to make facts.py python3 parsable

Commit:
move machine_id back to Facts class

robinro commented:
rebased and squashed and ready to be merged

Sometime in the future one should extend the refactor to clean out all TODOs and make sure each file under /etc is read/parsed only once, but without more testcases there is a high probability to break some rare case. The general problem is that there is often redundant but conflicting information (both etc/suse-release and /etc/os-release exist with different content). I'm happy to help with that though if someone feels adventurous :roller_coaster:

Generating testcases is easy, just run https://github.com/ansible/ansible/blob/692bf51fde67d8aa29c9928bd5719cf53a05154b/hacking/tests/gen_distribution_version_testcase.py on the test machine (assumes python and ansible working). Or collect the necessary data manually.

abadger commented:
Thanks @robinro!  Yeah -- I think it makes sense to go forth and make possibly breaking changes during the first part of 2.2 development.  Gives users the maximum amount of time to encounter the problem in devel and us to fix it.

Maybe send a message out to the ansible mailing list letting people know what we're doing here and to test, to run the script and contribute the generated test cases, etc.
